### PR TITLE
feat(c++): label filtering API, benchmarks, and examples

### PR DESCRIPTION
**Pull request by Elssky**
### Reason for this PR

Provides API for label filtering

### What changes are included in this PR?

New features about label--filtering vertices with specific labels
### Are these changes tested?

yes, and efficiency are shown in bechmark, correctness are shown in example
### Are there any user-facing changes?

yes, feature as above

### CODE REVIEW

File: cpp/src/graphar/high-level/graph_reader.cc [hunk 1]
```diff
@@ -94,6 +100,290 @@ Vertex::Vertex(IdType id,
...
+
+  for (int value : indices) {
+    indices64.push_back(static_cast<int64_t>(value));
+  }
+
+  return indices64;
+}
+
+Result<std::shared_ptr<VerticesCollection>>
+VerticesCollection::verticesWithLabel(
```
Comment (Elssky) [inline cpp/src/graphar/high-level/graph_reader.cc:268]: Single label filtering API

File: cpp/src/graphar/high-level/graph_reader.cc [hunk 2]
```diff
@@ -94,6 +100,290 @@ Vertex::Vertex(IdType id,
...
+  auto vertices_collection =
+      std::make_shared<VerticesCollection>(vertex_info, prefix);
+  vertices_collection->filtered_ids_ =
+      vertices_collection->filter({filter_label}).value();
+  vertices_collection->is_filtered_ = true;
+  return vertices_collection;
+}
+
+Result<std::shared_ptr<VerticesCollection>>
+VerticesCollection::verticesWithLabelbyAcero(
```
Comment (Elssky) [inline cpp/src/graphar/high-level/graph_reader.cc:283]: Single label filtering API（Acero）

File: cpp/src/graphar/high-level/graph_reader.cc [hunk 3]
```diff
@@ -94,6 +100,290 @@ Vertex::Vertex(IdType id,
...
+
+    new_vertices_collection->is_filtered_ = true;
+  }
+  new_vertices_collection->filtered_ids_ = filtered_ids;
+
+  return new_vertices_collection;
+}
+
+Result<std::shared_ptr<VerticesCollection>>
+VerticesCollection::verticesWithMultipleLabels(
```
Comment (Elssky) [inline cpp/src/graphar/high-level/graph_reader.cc:328]: Multi label filtering API

File: cpp/src/graphar/high-level/graph_reader.cc [hunk 4]
```diff
@@ -94,6 +100,290 @@ Vertex::Vertex(IdType id,
...
+  auto labels = vertex_info->GetLabels();
+  auto vertices_collection =
+      std::make_shared<VerticesCollection>(vertex_info, prefix);
+  vertices_collection->filtered_ids_ =
+      vertices_collection->filter_by_acero({filter_label}).value();
+  vertices_collection->is_filtered_ = true;
+  return vertices_collection;
+}
+
+Result<std::shared_ptr<VerticesCollection>>
```
Comment (Elssky) [inline cpp/src/graphar/high-level/graph_reader.cc:297]: Single label filtering API, input is VerticesCollection ( a result from another filtering)

File: cpp/src/graphar/high-level/graph_reader.cc [hunk 5]
```diff
@@ -94,6 +100,290 @@ Vertex::Vertex(IdType id,
...
+  auto vertices_collection =
+      std::make_shared<VerticesCollection>(vertex_info, prefix);
+  vertices_collection->filtered_ids_ =
+      vertices_collection->filter(filter_labels).value();
+  vertices_collection->is_filtered_ = true;
+  return vertices_collection;
+}
+
+Result<std::shared_ptr<VerticesCollection>>
+VerticesCollection::verticesWithMultipleLabelsbyAcero(
```
Comment (Elssky) [inline cpp/src/graphar/high-level/graph_reader.cc:343]: Multi label filtering API (Acero)

File: cpp/src/graphar/high-level/graph_reader.cc [hunk 6]
```diff
@@ -94,6 +100,290 @@ Vertex::Vertex(IdType id,
...
+  auto vertices_collection =
+      std::make_shared<VerticesCollection>(vertex_info, prefix);
+  vertices_collection->filtered_ids_ =
+      vertices_collection->filter_by_acero(filter_labels).value();
+  vertices_collection->is_filtered_ = true;
+  return vertices_collection;
+}
+
+Result<std::shared_ptr<VerticesCollection>>
+VerticesCollection::verticesWithMultipleLabels(
```
Comment (Elssky) [inline cpp/src/graphar/high-level/graph_reader.cc:358]: Multi label filtering API, input is VerticesCollection

File: cpp/src/graphar/high-level/graph_reader.h [hunk 7]
```diff
@@ -172,40 +179,80 @@ class VertexIter {
    * @param offset The current offset of the readers.
    */
   explicit VertexIter(const std::shared_ptr<VertexInfo>& vertex_info,
-                      const std::string& prefix, IdType offset) noexcept {
+                      const std::string& prefix, IdType offset,
+                      const std::vector<std::string>& labels,
+                      const bool& is_filtered = false,
+                      const std::vector<IdType>& filtered_ids = {}) noexcept {
+    if (!labels.empty()) {
```
Comment (Elssky) [inline cpp/src/graphar/high-level/graph_reader.h:186]: label_reader is used for reading label chunks

File: cpp/src/graphar/high-level/graph_reader.h [hunk 8]
```diff
@@ -172,40 +179,80 @@ class VertexIter {
...
+      labels_ = labels;
+      label_reader_ =
+          VertexPropertyArrowChunkReader(vertex_info, labels, prefix);
+    }
     for (const auto& pg : vertex_info->GetPropertyGroups()) {
       readers_.emplace_back(vertex_info, pg, prefix);
     }
+    is_filtered_ = is_filtered;
```
Comment (Elssky) [inline cpp/src/graphar/high-level/graph_reader.h:194]: is_filtered_ is used for mark the collection/iter is or not a result from filtering

File: cpp/src/graphar/high-level/graph_reader.h [hunk 9]
```diff
@@ -172,40 +179,80 @@ class VertexIter {
...
+    filtered_ids_ = filtered_ids;
```
Comment (Elssky) [inline cpp/src/graphar/high-level/graph_reader.h:195]: the vector of vertex ids after filtering

**Review by lixueclaire — APPROVED** [review]
LGTM！Thank you for your contribution~

File: cpp/src/graphar/high-level/graph_reader.cc [hunk 10]
```diff
@@ -94,6 +100,290 @@ Vertex::Vertex(IdType id,
...
+    if (it != labels_.end()) {
+      tested_label_ids.push_back(std::distance(labels_.begin(), it));
+    }
+  }
+  if (tested_label_ids.empty())
+    return Status::KeyError(
+        "query label"
+        " does not exist in the vertex.");
+
+  uint64_t* bitmap = new uint64_t[TOT_ROWS_NUM / 64 + 1];
```
Comment (yangxk1) [inline cpp/src/graphar/high-level/graph_reader.cc:173]: I'm wondering why memory is allocated for the bitmap here — is it for security reasons or another purpose?
In the `read_parquet_file_and_get_valid_indices` method, the `query_type` argument is always passed as `QUERY_TYPE::INDEX`. It seems that the bitmap is never actually used.

Comment (yangxk1) [reply cpp/src/graphar/high-level/graph_reader.cc:173]: @Elssky

**Pull request merged**